### PR TITLE
[packagist] api v2 support

### LINKED ISSUE #7020: Packagist badge cannot handle composer v2-only packages
### Are you experiencing an issue with...

shields.io

### 🐞 Description

Packagist (the PHP dependency management SaaS) deprecated their v1 API some time ago, so new packages are only being added to their v2 API.

The shields site uses v1 of the API, so it is not able to retrieve version numbers for newer packages.

### 🔗 Link to the badge

New package, not working: ecoapm/libyear

Old package, works fine: symfony/symfony

### 💡 Possible Solution

There are a couple possible options:
1. update `BasePackagistService.fetch()` to use the new API and modify `transform()` accordingly
1. update `PackagistVersion.handle()` to fallback to the JSON API

I would think the first would be preferrable, from a future maintenance perspective, so I'll get a PR ready for that tomorrow.

### PR DESCRIPTION
**Pull request by yookoala**
Rebased #6508 on the current master branch. An attempt to fix #7020.

I have a packagist v2 only package ([this one](https://packagist.org/packages/monogatari-php/remote-storage)). The v1 API doesn't really show any information about my package. I'd want to fix this so badges for my package works normally. Please let me know what I need to do to get this merged.

---

Update: Seems the badge for my package was affected by the [reduced update rate of v1 Packagist API](https://blog.packagist.com/deprecating-composer-1-support/#reduced-v1-metadata-api-update-rate-starting-in-may-2021). It works for me now. But I'd still like to contribute and fix this for long run.

### CODE REVIEW

**Comment by chris48s** [comment]
Thanks for picking this up. #6508 was most of the way there and it would be good to get this over the line but it had fallen off the radar.

It has been quite a while since I looked at #6508 but having got back into it and looked over this, I've pushed a few more commits with my suggested changes to get this over the line.

Maybe another maintainer could give this a once over so we can get this one resolved.

---

In response to https://github.com/badges/shields/pull/6508#discussion_r800121184 (I'll reply here to keep the conversation in this thread rather than split across two PRs):

The logic in `packagist-version.service.js` should only be dealing with tagged releases and the code doesn't call the `~dev` endpoint, so we don't need the code for dealing with branch aliases.

**Comment by yookoala** [comment]
> In response to [#6508 (comment)](https://github.com/badges/shields/pull/6508#discussion_r800121184) (I'll reply here to keep the conversation in this thread rather than split across two PRs):
> 
> The logic in `packagist-version.service.js` should only be dealing with tagged releases and the code doesn't call the `~dev` endpoint, so we don't need the code for dealing with branch aliases.

I think we need some clarification before carrying on the discussion.

If I read the code correct:

1. `PackagistVersion` never referenced `BasePackagistService.fetchDev`. I suppose that means it's not calling the `~dev.json` endpoint, correct?
2. `Packagist.transform` seems to be map all the `version` attribute (i.e. the git tag) from the packages array items. Since I suppose it is using the ordinary p2 endpoint, the returned packages are all *tagged* commits, correct?
3. If `includePrereleases` is `false`, the *tag* strings are filtered by the function `isStable()` [here](https://github.com/badges/shields/blob/e3854098d629e9c390196df0c0fd2fd5795b6bca/services/php-version.js#L167-L177) from `php-version.js` is checking the pattern of the tag. If the tag ends with nothing (modifierLevel 3) or a `patch*` (modifierLevel 4), then the tag is consider a stable version tag. If the tag ends with `alpha*`, `beta*`, `dev*` or anything else, the provided version string is considered "not stable". This seems logical to me.
4. As mentioned in (2), no matter if `includePrereleases` is `true` or `false`, the filter target seems to be only the tagged release. And as described in (3), the `Packagist.transform` method seems to be just checking if the *tag* string pattern matches that of release versions. If both observations are correct, then nothing here involves development branch.

The logics here looks like it wants to fulfil the contract of the method `handle`, which has an `includePrereleases` parameter. The approach taken to fulfil that contract seems reasonable, too.

**Comment by chris48s** [comment]
> If I read the code correct:
> 
> 1. `PackagistVersion` never referenced `BasePackagistService.fetchDev`. I suppose that means it's not calling the `~dev.json` endpoint, correct?
> 2. `Packagist.transform` seems to be map all the `version` attribute (i.e. the git tag) from the packages array items. Since I suppose it is using the ordinary p2 endpoint, the returned packages are all *tagged* commits, correct?
> 3. If `includePrereleases` is `false`, the *tag* strings are filtered by the function `isStable()` [here](https://github.com/badges/shields/blob/e3854098d629e9c390196df0c0fd2fd5795b6bca/services/php-version.js#L167-L177) from `php-version.js` is checking the pattern of the tag. If the tag ends with nothing (modifierLevel 3) or a `patch*` (modifierLevel 4), then the tag is consider a stable version tag. If the tag ends with `alpha*`, `beta*`, `dev*` or anything else, the provided version string is considered "not stable". This seems logical to me.
> 4. As mentioned in (2), no matter if `includePrereleases` is `true` or `false`, the filter target seems to be only the tagged release. And as described in (3), the `Packagist.transform` method seems to be just checking if the *tag* string pattern matches that of release versions. If both observations are correct, then nothing here involves development branch.

Yes I'd agree with that summary :+1:

**Comment by yookoala** [comment]
@chris48s: I've modified `BasePackagistService.findLatestVersion` to accept an optional `includePrereleases` parameter. Then I simplified the logics in `PackagistVersion.handle` to use `BasePackagistService.findLatestVersion` instead of its own transform method. Please check if this is what you wanted.

**Comment by yookoala** [comment]
Please 

> @chris48s: I've modified `BasePackagistService.findLatestVersion` to accept an optional `includePrereleases` parameter. Then I simplified the logics in `PackagistVersion.handle` to use `BasePackagistService.findLatestVersion` instead of its own transform method. Please check if this is what you wanted.

Scrap that. Need to make sure `PackageLicense` works here but I'm not sure about it.

Need further checking.

**Comment by yookoala** [comment]
@chris48s: OK. Here's what I found:

1. In the v2 API, only the first item in the package array has the "version" attribute.
2. Since not all items has the "version" attribute, there is no guarantee if the latest semver has a "license" attribute. That is only by chance.

Here are the different needs to the license and latest version service:
1. License service: get the first item in the API versions array and return the version.
2. Latest version service: get the most advanced semver of "version" attribute from API versions array (with an option to select pre-release).

These 2 logics are not quite compatible. I don't think its suitable to refactor them into one.

**Comment by yookoala** [comment]
@chris48s: I've modified `BasePackagistService.findLatestVersion` to accept an optional `includePrereleases` parameter. Then I simplified the logics in `PackagistVersion.handle` to use `BasePackagistService.findLatestVersion` instead of its own transform method.

Also did some rename and documentation of the existing methods.

Please check if this is what you asked for.

**Comment by yookoala** [comment]
@chris48s: Should have fixed everything for now. Please let me know if there is anything else needed to change before merging.

**Comment by chris48s** [comment]
Refactoring looks broadly sensible  and thanks for adding more tests but it looks like one of the changes in these last 2 commits is causing one of the service tests to fail.
`npm run test:services -- --only="packagist"`
Can you have a look at this? Thanks

**Comment by yookoala** [comment]
@chris48s: Added another commit to fix the issue. Also refactored the NotFound logics into findLatestRelease to streamline the code.

**Comment by yookoala** [comment]
> I don't know why the CI build is not triggering when you push, but there is now a test failing with
> 
> ```
> AssertionError: expected promise to be rejected with an error including 'version requirement not found' but it was fulfilled with { phpVersion: '__unset' }
> ```

@chris48s: Since I'm a first time committer, workflow will need approval before triggering by me:
2022-02-11 09-14-36 的螢幕擷圖

I've changed the newly written test to reflect the actual API behaviour. It was wrongly presume that the minifier is doing thing recursively.

**Comment by yookoala** [comment]
@calebcartwright This is quite annoying 😅

**Comment by calebcartwright** [comment]
Don't worry about keeping the branch in sync for now because it's the day when the bots bump all our dependencies so likely to need frequent updates.

Once we're ready to move forward after the review is completed we can do one last juggle (which may require closing and reopening the PR again)

**Review by chris48s — DISMISSED** [review]
I reckon this is good to go, but as I've also pushed a bunch of commits to this branch I'm going to leave it open for the moment in case another maintainer wants to weigh in on it.. If not, let me know and I'll just merge it

**Comment by calebcartwright** [comment]
I can probably set aside some today later today or tomorrow if another pair of eyes would be helpful, though readily confess much of the upstream context and specifics have gone over my head

**Comment by yookoala** [comment]
A bit of context might help you read the PR more easily. The major change from Packagist API v1 to v2:

## Major Changes in API

1. The v1 endpoint shows all the packages of the same user in the result. v2 endpoint only show the relevant package.
2. The package node in v1 is a key-value object of versions. With git tags (e.g. v3.7.0) to be the keys. In v2, the package node is an array of versions ordered by either release date or the version tag ([the documentation](https://packagist.org/apidoc#get-package-data) didn't specified).
3. The versions array in v2 is minified by the [MetadataMinifier](https://github.com/composer/metadata-minifier/blob/main/src/MetadataMinifier.php). In the package versions array, a missing key in later item(s) means the value(s) are unchanged from the previous one. If a key is to be removed, the key will be set to an explicit string value `"__unset"`.

## Key Changes to the Service Here

1. Update to use v2 endpoint.
2. Update schema to reflect the current API changes.
3. Add a `expandPackageVersions` method in the base class to expand what was minified by [MetadataMinifier](https://github.com/composer/metadata-minifier/blob/main/src/MetadataMinifier.php).
4. Changed from reading information of the "default branch" to reading information of the latest release version.

## Examples

Example of v1 and v2 API endpoint:
* v1: https://repo.packagist.org/p/phpunit/phpunit.json
  ```json
  {
    "packages": {
      "camdram/camdram": {...},
      "irstea/phpunit-shim": {...},
      "jworman/phpunit": {...},
      "lipemat/wp-unit": {...},
      "nanasess/phpunit48": {...},
      "phibi/pumore": {...},
      "php5friends/phpunit48": {...},
      "phpunit/phpunit": {
        "3.7.0": {
          "name": "phpunit/phpunit",
          "description": "The PHP Unit Testing framework.",
          "keywords": ["testing", "phpunit", "xunit"],
          "homepage": "http://www.phpunit.de/",
          "version": "3.7.0",
          "version_normalized": "3.7.0.0",
          "license": ["BSD-3-Clause"],
          "authors": [
            {
              "name": "Sebastian Bergmann",
              "email": "sebastian@phpunit.de",
              "role": "lead"
            }
          ],
          "source": {
            "url": "https://github.com/sebastianbergmann/phpunit.git",
            "type": "git",
            "reference": "f6a4872c01bab898bce4f9e211e437533f248d43"
          },
          "dist": {
            "url": "https://api.github.com/repos/sebastianbergmann/phpunit/zipball/f6a4872c01bab898bce4f9e211e437533f248d43",
            "type": "zip",
            "shasum": "",
            "reference": "f6a4872c01bab898bce4f9e211e437533f248d43"
          },
          "type": "library",
          "time": "2012-09-19T05:07:48+00:00",
          "autoload": { "files": ["PHPUnit/Autoload.php"] },
          "extra": { "branch-alias": { "dev-master": "3.7.x-dev" } },
          "include-path": ["", "../../symfony/yaml/"],
          "bin": ["composer/bin/phpunit"],
          "require": {
            "php": ">=5.3.3",
            "phpunit/php-file-iterator": ">=1.3.1@stable",
            "phpunit/php-text-template": ">=1.1.1@stable",
            "phpunit/php-code-coverage": ">=1.2.0@stable",
            "phpunit/php-timer": ">=1.0.2@stable",
            "phpunit/phpunit-mock-objects": ">=1.2.0@stable",
            "symfony/yaml": ">=2.1.0@stable",
            "ext-dom": "*",
            "ext-pcre": "*",
            "ext-reflection": "*",
            "ext-spl": "*"
          },
          "suggest": {
            "phpunit/php-invoker": ">=1.1.0@stable",
            "ext-json": "*",
            "ext-simplexml": "*",
            "ext-tokenizer": "*"
          },
          "uid": 19572
        },
        "3.7.1": {...},
        ...
      },
      "sminnee/phpunit": {...},
      "zf1s/phpunit": {...}
    }
  }

  ```
* v2: https://repo.packagist.org/p2/phpunit/phpunit.json
  ```json
  {
    "minified": "composer/2.0",
    "packages": {
      "phpunit/phpunit": [
        {
          "name": "phpunit/phpunit",
          "description": "The PHP Unit Testing framework.",
          "keywords": ["testing", "phpunit", "xunit"],
          "homepage": "https://phpunit.de/",
          "version": "9.5.13",
          "version_normalized": "9.5.13.0",
          "license": ["BSD-3-Clause"],
          "authors": [
            {
              "name": "Sebastian Bergmann",
              "email": "sebastian@phpunit.de",
              "role": "lead"
            }
          ],
          "source": {
            "url": "https://github.com/sebastianbergmann/phpunit.git",
            "type": "git",
            "reference": "597cb647654ede35e43b137926dfdfef0fb11743"
          },
          "dist": {
            "url": "https://api.github.com/repos/sebastianbergmann/phpunit/zipball/597cb647654ede35e43b137926dfdfef0fb11743",
            "type": "zip",
            "shasum": "",
            "reference": "597cb647654ede35e43b137926dfdfef0fb11743"
          },
          "type": "library",
          "funding": [
            { "url": "https://phpunit.de/sponsors.html", "type": "custom" },
            { "url": "https://github.com/sebastianbergmann", "type": "github" }
          ],
          "time": "2022-01-24T07:33:35+00:00",
          "autoload": {
            "files": ["src/Framework/Assert/Functions.php"],
            "classmap": ["src/"]
          },
          "extra": { "branch-alias": { "dev-master": "9.5-dev" } },
          "bin": ["phpunit"],
          "require": {
            "php": ">=7.3",
            "ext-dom": "*",
            "ext-json": "*",
            "ext-libxml": "*",
            "ext-mbstring": "*",
            "ext-xml": "*",
            "ext-xmlwriter": "*",
            "doctrine/instantiator": "^1.3.1",
            "myclabs/deep-copy": "^1.10.1",
            "phar-io/manifest": "^2.0.3",
            "phar-io/version": "^3.0.2",
            "phpspec/prophecy": "^1.12.1",
            "phpunit/php-code-coverage": "^9.2.7",
            "phpunit/php-file-iterator": "^3.0.5",
            "phpunit/php-invoker": "^3.1.1",
            "phpunit/php-text-template": "^2.0.3",
            "phpunit/php-timer": "^5.0.2",
            "sebastian/cli-parser": "^1.0.1",
            "sebastian/code-unit": "^1.0.6",
            "sebastian/comparator": "^4.0.5",
            "sebastian/diff": "^4.0.3",
            "sebastian/environment": "^5.1.3",
            "sebastian/exporter": "^4.0.3",
            "sebastian/global-state": "^5.0.1",
            "sebastian/object-enumerator": "^4.0.3",
            "sebastian/resource-operations": "^3.0.3",
            "sebastian/type": "^2.3.4",
            "sebastian/version": "^3.0.2"
          },
          "require-dev": { "ext-pdo": "*", "phpspec/prophecy-phpunit": "^2.0.1" },
          "suggest": { "ext-soap": "*", "ext-xdebug": "*" },
          "support": {
            "issues": "https://github.com/sebastianbergmann/phpunit/issues",
            "source": "https://github.com/sebastianbergmann/phpunit/tree/9.5.13"
          }
        },
        {
          "version": "9.5.12",
          "version_normalized": "9.5.12.0",
          "source": {
            "url": "https://github.com/sebastianbergmann/phpunit.git",
            "type": "git",
            "reference": "93d4bf4c37aec6384bb9e5d390d9049a463a7256"
          },
          "dist": {
            "url": "https://api.github.com/repos/sebastianbergmann/phpunit/zipball/93d4bf4c37aec6384bb9e5d390d9049a463a7256",
            "type": "zip",
            "shasum": "",
            "reference": "93d4bf4c37aec6384bb9e5d390d9049a463a7256"
          },
          "time": "2022-01-21T05:54:47+00:00",
          "support": {
            "issues": "https://github.com/sebastianbergmann/phpunit/issues",
            "source": "https://github.com/sebastianbergmann/phpunit/tree/9.5.12"
          }
        },
        {
          "version": "9.5.11",
          "version_normalized": "9.5.11.0",
          ...
        }
        ...
      ]
    }
  }
  ```

**Comment by calebcartwright** [comment]
Thank you for the summary @yookoala! I obviously didn't get to this over the weekend but it's top of my list next time I've a moment to sit and dig in

File: services/packagist/packagist-base.js
Comment (calebcartwright) [inline services/packagist/packagist-base.js:146]: Was going to ask if it'd be worth adding some unit tests for this function since the logic, though not terribly complex, isn't exactly trivial either.

I see that we're indirectly testing it with tests that targetting some of the child classes. That works for me, at least as long as we don't end up duplicating tests down the road

File: services/packagist/packagist-php-version.service.js
Comment (calebcartwright) [inline services/packagist/packagist-php-version.service.js:97]: I believe my mental context is fresh enough at the moment to follow this, but if anyone is receptive and willing to do so in a follow up PR, I think a brief inline comment summarizing the what/why would be helpful for future readers (likely myself included)

**Review by calebcartwright — APPROVED** [review]
Couple thoughts left inline, but largely as considerations for small potential future work

**Comment by yookoala** [comment]
@Seldaek: Did some review on the behaviour of Packagist, the default page of a package seems shows:

1. The latest stable version. If not,
2. The latest version on the default branch.

Can you please confirm this. I'd like to rewrite shields.io version search with the default behaviour of Packagist site.

File: services/packagist/packagist-base.js
Comment (yookoala) [reply services/packagist/packagist-base.js]: Add more testing on the latest commit. Please check.

**Comment by Seldaek** [comment]
> I did some review on the behaviour of Packagist, the default page of a package seems shows:
> 
> 1. The latest stable version. If not,
> 2. The latest version on the default branch.

Yes that seems accurate, per https://github.com/composer/packagist/blob/7493c36087514fbe269bc623edba9b8d6d1c96eb/src/Controller/PackageController.php#L570-L583 and https://github.com/composer/packagist/blob/7493c36087514fbe269bc623edba9b8d6d1c96eb/src/Entity/Package.php#L871-L874

The one slight inaccuracy is if the default branch is numeric and for example `2.x` but you have a `3.x` branch, then I believe it would still show 3.x-dev even tho that's not the default branch, but that only happens if you have no tagged version which is very rare anyway.

**Comment by chris48s** [comment]
I've kind of lost track of where we are with this PR but I'd really like to try and get this work landed one way or another.

The branch has been force-pushed a bunch of times so I've totally lost the thread of what changed since last time I reviewed it.

@yookoala - are you waiting on us to review this, or are we waiting for you to update it?

**Comment by yookoala** [comment]
I believe I made all changes that are asked by you and other reviewers. Please kindly check if any other amendment you might want me to do.

**Comment by chris48s** [comment]
Just having a cursory look at the build it is failing with `InvalidService: Expected /usr/src/app/services/packagist/packagist-license.service.js to export a service or a collection of services; one of them was license not found`. Can you have a look at this and make sure what is on this branch is definitely ready for review and I will try and find some time to get back into it next week.

**Comment by yookoala** [comment]
> Just having a cursory look at the build it is failing with `InvalidService: Expected /usr/src/app/services/packagist/packagist-license.service.js to export a service or a collection of services; one of them was license not found`. Can you have a look at this and make sure what is on this branch is definitely ready for review and I will try and find some time to get back into it next week.

Should be fixed now. I've also re-run `npm run test:core -- --only="packagist"`. Every tests are passed.

Please let me know what to check next, if any. Thanks.

**Comment by chris48s** [comment]
Thanks for your continued work on this but I tried re-running the build with the latest changes. There are still failing tests:

```
PackagistLicense should return the license of the most recent release
should return the license of the most recent release
/home/circleci/project/services/packagist/packagist-license.spec.js

TypeError: PackagistLicense.prototype.getLicense is not a function
    at Context.<anonymous> (file:///home/circleci/project/services/packagist/packagist-license.spec.js:30:39)
    at processImmediate (node:internal/timers:464:21)

PackagistLicense should return the license of the most recent stable release
should return the license of the most recent stable release
/home/circleci/project/services/packagist/packagist-license.spec.js

TypeError: PackagistLicense.prototype.getLicense is not a function
    at Context.<anonymous> (file:///home/circleci/project/services/packagist/packagist-license.spec.js:55:39)
    at processImmediate (node:internal/timers:464:21)

PackagistLicense should return the license of the most recent pre-release if no stable releases
should return the license of the most recent pre-release if no stable releases
/home/circleci/project/services/packagist/packagist-license.spec.js

TypeError: PackagistLicense.prototype.getLicense is not a function
    at Context.<anonymous> (file:///home/circleci/project/services/packagist/packagist-license.spec.js:80:39)
    at processImmediate (node:internal/timers:464:21)

PackagistLicense should throw NotFound when license key not in response
should throw NotFound when license key not in response
/home/circleci/project/services/packagist/packagist-license.spec.js

AssertionError: expected [Function] to throw 'NotFound' but 'TypeError: PackagistLicense.prototype…' was thrown
    at Context.<anonymous> (file:///home/circleci/project/services/packagist/packagist-license.spec.js:104:16)
    at processImmediate (node:internal/timers:464:21)

PackagistPhpVersion should return PHP version for the default branch
should return PHP version for the default branch
/home/circleci/project/services/packagist/packagist-php-version.spec.js

AssertionError: expected '^8.1' to equal '^7.4 || 8'
    at Context.<anonymous> (file:///home/circleci/project/services/packagist/packagist-php-version.spec.js:124:13)
    at processImmediate (node:internal/timers:464:21)
```

It seemed like we were close to getting this merged at one point. I can see you are trying to help, but it also feels like we're going round in circles a bit here. I am aware that the fact builds are not running on this branch for some reason is making it harder for both of us. I'm also finding the changes from e5b31cd onwards quite hard to understand. There is a lot of tangential refactoring mixed in with functional changes and I'm finding it difficult to follow the thread of where the behaviour of the code is changing intentionally, and where it is changing accidentally. I think you are too tbh - there are several points where this PR has been "ready" with tests still failing.

This is the second time someone has tried to pick up this issue and I'd really like to try and get #7020 closed. I'd like this not to fall through the cracks again. Here's an idea for a way forward:

If we take the code on this branch up to b164565 , is what's there fundamentally broken? i.e: if we merge the commit range [2a87125 - b164565](https://github.com/badges/shields/compare/master...yookoala:b164565b3a15f0656ed8e17a26dc2636b32747ab) to master, are we actively introducing bugs? If the answer to that is no, I'm going to suggest we pick those commits to a new branch, merge that and then if you have some additional functional improvements or refactoring PRs you want to make to the packagist suite of badges you can make those changes in one or more follow up PRs which are more clearly scoped. This will be easier for both of us to reason about (and hopefully the builds will run automatically on a new PR). I accept that it is not always possible to completely decouple refactoring from functional changes and there are improvements that can be made to this code, but if we can break this down into smaller clearer objectives rather than trying to boil the ocean in one PR we're much more likely to achieve success.

**Comment by yookoala** [comment]
I'd agree.

Also, what is the command to run all tests for packagist?

**Comment by calebcartwright** [comment]
> Also, what is the command to run all tests for packagist?

`npm run test:services -- --only=packagist`

**Comment by chris48s** [comment]
> > Also, what is the command to run all tests for packagist?
>
> `npm run test:services -- --only=packagist`

Sort of.

There are two types of tests relevant to this PR:

### Unit tests
These are in files ending `.spec.js` and we run all of them when we run `npm test` (you can actually run `npm run test:core` to run these if you want - this is faster because it skips the frontend tests and the linter checks). The unit tests are quick to run and don't call out to any external services. These generally test quite specific functionality.

### Service tests
These are in files ending `.tester.js`. These are a type of integration test. They're a bit higher level and they actually call out to external services. This is unusual practice in testing but we do this because one of the purposes of service tests is to notify us if a badge has broken due to an upstream API change. Because these tests call out to external APIs they are slower to run and can sometimes be flaky which is why we don't run them all on every PR or when we call `npm test` (but we do have a scheduled job that runs the full suite of service tests once a day). To run the packagist service tests locally you run `npm run test:services -- --only=packagist` (as @calebcartwright  rightly says). Maybe one of the things that might have happened is you're modifying the unit tests (`.spec.js` files) but then running the service tests with `npm run test:services -- --only=packagist` and seeing them pass, or vice versa?

Hopefully that helps.

---

In any case, I have pulled out 2a87125 - b164565 into another PR over at [#num]. Lets get that one merged first. I will close this PR. Then if you want to do some follow-up PRs with functional tweaks, additional testing or structural improvements incorporating some of the concepts from this branch then we can look at them. Splitting that into a few smaller PRs with a clear objective would be really helpful.

**Comment by calebcartwright** [comment]
Yeah good shout. Was largely doing a drive by response from the notification inbox and assumed we were talking about service tests; missed the context of the failing tests in the core suite

**Pull request closed without merging**
